x-pack/legacy/plugins/lens/public/datatable_visualization_plugin/visualization.tsx

```diff
@@ -6,150 +6,91 @@
-                    iconType="trash"
-                    onClick={() => {
-                      datasource.removeColumnInTableSpec(id);
-                      const newColumns = [...state.columns];
-                      newColumns.splice(index, 1);
-                      setState({
-                        ...state,
-                        columns: newColumns,
-                      });
-                    }}
-                    aria-label={i18n.translate('xpack.lens.datatable.removeColumnAriaLabel', {
-                      defaultMessage: 'Remove',
-                    })}
-                  />
-                </EuiFlexItem>
-              </EuiFlexGroup>
-            </EuiFormRow>
-          </>
-        );
-      })}
+function updateColumns(
+  state: DatatableVisualizationState,
+  layer: LayerState,
+  fn: (columns: string[]) => string[]
+) {
+  const columns = fn(layer.columns);
+  const updatedLayer = { ...layer, columns };
+  const layers = state.layers.map(l => (l.layerId === layer.layerId ? updatedLayer : l));
+  return { ...state, layers };
+}
 
-      <div>
-        <EuiButton
-          data-test-subj="lnsDatatable_dimensionPanel_add"
-          onClick={() => {
-            const newColumns = [...state.columns];
-            newColumns.push({
-              id: generateId(),
-              label: '',
-            });
-            setState({
-              ...state,
-              columns: newColumns,
-            });
-          }}
-          iconType="plusInCircle"
-        />
-      </div>
-    </EuiForm>
+export function DataTableLayer({
+  layer,
+  frame,
+  state,
+  setState,
+  dragDropContext,
+}: { layer: LayerState } & VisualizationProps<DatatableVisualizationState>) {
+  const datasource = frame.datasourceLayers[layer.layerId];
+  return (
+    <EuiFormRow
+      label={i18n.translate('xpack.lens.datatable.columns', { defaultMessage: 'Columns' })}
+    >
+      <MultiColumnEditor
+        accessors={layer.columns}
+        datasource={datasource}
+        dragDropContext={dragDropContext}
+        filterOperations={() => true}
+        layerId={layer.layerId}
+        onAdd={() => setState(updateColumns(state, layer, columns => [...columns, generateId()]))}
+        onRemove={column =>
+          setState(updateColumns(state, layer, columns => columns.filter(c => c !== column)))
+        }
+        testSubj="datatable_columns"
+        data-test-subj="datatable_multicolumnEditor"
+      />
+    </EuiFormRow>
   );
 }
 
 export const datatableVisualization: Visualization<
   DatatableVisualizationState,
   DatatableVisualizationState
 > = {
-  initialize(datasource, state) {
+  initialize(frame, state) {
+    const layerId = Object.keys(frame.datasourceLayers)[0] || frame.addNewLayer();
     return (
       state || {
-        columns: [
-          {
-            id: generateId(),
-            label: '',
-          },
-        ],
+        layers: [newLayerState(layerId)],
```

Comment:
It seems like something doesn't work correctly here - if I switch to the datatable vis without changing anything in the xy vis, it doesn't show an empty column to configure
<img width="432" alt="Screenshot 2019-07-23 at 18 19 15" src="https://user-images.githubusercontent.com/1508364/61729049-b8d19580-ad76-11e9-8253-bebee5da41c4.png">

Reply:
Hm. This is one of the messes that I don't like. Right now, the responsibility for ensuring an empty trailing column belongs to the multicolumn editor UI component. It *really* belongs in a reducer or state management function. For now, I'll update the broken logic, but long term, I think we should come up with a better way (reducer?) of centralizing state changes / sanitization for each visualization.

Reply:
This makes sense, maybe we can collect helper functions for common state handling things a lot of visualizations are going to require.